### PR TITLE
Migrate analyzer caches to IndexVersion

### PR DESCRIPTION
**Pull request by thecoop**
Migrate pre-built analyzers and caches to use `IndexVersion` rather than `Version`.

This changes some APIs that are used by plugins & other external code.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeVersionAllocationDecider.java [hunk 1]
```diff
@@ -118,22 +118,22 @@ private static Decision isVersionCompatible(
         final RoutingNode target,
         final RoutingAllocation allocation
     ) {
-        if (target.node().getVersion().onOrAfter(recoverySource.version().toVersion())) {
+        if (target.node().getMaxIndexVersion().onOrAfter(recoverySource.version())) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeVersionAllocationDecider.java:121]: I think this is unrelated to analysis changes?

**Review by romseygeek — APPROVED** [review]
LGTM.  We don't have any guarantees on plugin compatibility, but it might be worth adding something to the migration docs about the changes here as analysis plugins are some of the more popular third-party plugins out there.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeVersionAllocationDecider.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeVersionAllocationDecider.java:121]: Oh yes, a stray unrelated change snuck in
